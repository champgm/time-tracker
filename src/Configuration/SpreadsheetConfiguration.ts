export interface SpreadsheetConfiguration {
  spreadsheetId: string,
  sheetId:number,
}