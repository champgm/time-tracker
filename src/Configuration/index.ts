import { googleToken } from "./GoogleToken";
import { spreadsheetConfiguration } from "./Spreadsheet";

export interface Configuration {
  spreadsheetId: string,
  sheetId:number,
  googleAuth: {
    sheetsEndpoint: string,
    tokenEndpoint: string,
    tokenRequestBody: {
      grant_type: string,
      refresh_token: string,
      client_id: string,
      client_secret: string,
    },
  };
}

export const configuration: Configuration = {
  ...spreadsheetConfiguration,
  googleAuth: {
    sheetsEndpoint: "https://sheets.googleapis.com/v4/spreadsheets",
    tokenEndpoint: "https://www.googleapis.com/oauth2/v4/token",
    tokenRequestBody: googleToken,
  },
};
