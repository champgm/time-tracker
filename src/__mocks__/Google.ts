
export const sendEmail = jest.fn(() => {
  console.log(`Mocked 'sendEmail' function called.`);
});
