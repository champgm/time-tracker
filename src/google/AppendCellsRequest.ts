import { RowData } from "./RowData";

export interface AppendCellsRequest {
  sheetId: number,
  rows: RowData[],
  fields: string
}