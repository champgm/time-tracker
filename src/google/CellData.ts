import { ExtendedValue } from './ExtendedValue';

export interface CellData {
  userEnteredValue: ExtendedValue
}