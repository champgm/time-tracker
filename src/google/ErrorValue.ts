
export interface ErrorValue {
  "type": string,
  "message": string
}