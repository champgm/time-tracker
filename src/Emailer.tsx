import _ from "lodash";
import React from "react";
import {
  Dimensions,
  StyleSheet,
  TextInput,
  TextStyle,
  ToastAndroid,
  View,
} from "react-native";
import AwesomeButton from "react-native-really-awesome-button";
import v4 from "uuid/v4";

import { createBasesFromColor, rgb, rgbStrings as bases } from "solarizer";
import { configuration, Recipient } from "./Configuration";
import { sendEmail } from "./Google";

const blue = createBasesFromColor(rgb.blue, "base01");
const red = createBasesFromColor(rgb.red, "base01");
const green = createBasesFromColor(rgb.green, "base01");

export interface Props { }

interface State {
  recipients: { [name: string]: Recipient };
  subject?: string;
  body?: string;
  sending?: boolean;
}

export class Emailer extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = { recipients: {} };
    configuration.recipients.forEach((recipient) => {
      this.state.recipients[recipient.name] = recipient;
      if (recipient.name === configuration.defaultRecipient) {
        this.state.recipients[recipient.name].selected = true;
      }
    });
  }

  toggleRecipientSelection(recipientName: string) {
    const recipient = this.state.recipients[recipientName];
    recipient.selected = !recipient.selected;
    this.setState(this.state);
  }

  resetState() {
    this.setState({
      subject: undefined,
      body: undefined,
      sending: undefined,
    });
  }

  async sendEmail() {
    this.setState({ sending: true });
    const recipients = Object.keys(this.state.recipients).map((recipientName) => this.state.recipients[recipientName]);
    const addresses = recipients.filter((recipient) => recipient.selected).map((recipient) => recipient.email);
    const errors: any[] = [];
    for (const address of addresses) {
      try {
        ToastAndroid.showWithGravity(
          "Sending email(s)...",
          ToastAndroid.SHORT,
          ToastAndroid.TOP,
        );
      } catch (error) {
        // console.log(`Could not show sending toast.`);
      }
      try {
        await sendEmail(
          configuration.sender,
          address,
          this.state.subject,
          this.state.body);
      } catch (error) {
        errors.push(error);
      }
    }

    if (errors.length > 0) {
      try {
        console.log(`${JSON.stringify(errors, null, 2)}`);
        ToastAndroid.showWithGravity(
          JSON.stringify(errors, null, 2),
          ToastAndroid.LONG,
          ToastAndroid.TOP,
        );
      } catch (error) {
        // console.log(`Could not show errors toast, error: ${error}`);
        // console.log(`Errors: ${errors}`);
      }
    } else {
      try {
        ToastAndroid.showWithGravity(
          "Successfully sent email(s)",
          ToastAndroid.SHORT,
          ToastAndroid.TOP,
        );
      } catch (error) {
        // console.log(`Could not show success toast: ${error}`);
      }
    }
    this.resetState();
  }

  render() {
    const emailButtons = Object.keys(this.state.recipients).map((recipientName) => {
      const recipient = this.state.recipients[recipientName];
      return <AwesomeButton
        key={this.state.recipients[recipientName].email}
        onPress={() => this.toggleRecipientSelection(recipientName)}
        accessibilityLabel={recipientName}
        backgroundColor={recipient.selected ? blue.base01 : bases.base01}
        backgroundActive={recipient.selected ? blue.base02 : bases.base02}
        backgroundDarker={recipient.selected ? blue.base03 : bases.base03}
        disabled={this.state.sending}
      >
        {` ${recipientName} `}
      </AwesomeButton>;
    });
    return (
      <View style={styles.root} >
        <View style={styles.emailForm} >
          <View style={styles.recipientButtonGroup}>{emailButtons}</View>
          <View>
            <TextInput
              // Subject Field
              onChangeText={(subject) => this.setState({ subject })}
              style={styles.subject}
              value={this.state.subject}
              autoFocus={true}
              testID="cd7b46a4-2d81-47bf-abc1-7142aba8a7b0"
            />
          </View>
          <View>
            <TextInput
              // Body Field
              multiline={true}
              numberOfLines={8}
              onChangeText={(body) => this.setState({ body })}
              style={styles.body}
              value={this.state.body}
              testID="6b50a3ac-a102-4150-823c-e20d44f0c84d"
            />
          </View>
          <View style={styles.actionButtonGroup}>
            <AwesomeButton
              // Clear button
              key={v4()}
              onPress={() => this.resetState()}
              accessibilityLabel="Clear"
              width={width / 3}
              backgroundColor={this.state.sending ? bases.base01 : red.base01}
              backgroundActive={this.state.sending ? bases.base02 : red.base02}
              backgroundDarker={this.state.sending ? bases.base03 : red.base03}
              disabled={this.state.sending}
            >
              Clear
            </AwesomeButton>
            <AwesomeButton
              // Submit Button
              key={v4()}
              onPress={() => this.sendEmail()}
              accessibilityLabel="Send"
              testID="2d8395f6-03a5-4c61-9c3b-595143aec8bf"
              width={width / 3}
              backgroundColor={this.state.sending ? bases.base01 : green.base01}
              backgroundActive={this.state.sending ? bases.base02 : green.base02}
              backgroundDarker={this.state.sending ? bases.base03 : green.base03}
              disabled={this.state.sending}
            >
              Send
            </AwesomeButton>
          </View>
        </View>
      </View >);
  }
}

const width = Dimensions.get("window").width * .7;
const marginTop = Dimensions.get("window").height * .03;
const textInputBorder: TextStyle = {
  borderColor: bases.base01,
  borderRadius: 5,
  borderWidth: 1,
};
const styles = StyleSheet.create({
  actionButtonGroup: {
    flexDirection: "row",
    justifyContent: "space-between",
    marginTop,
    width,
  },
  body: {
    paddingHorizontal: Dimensions.get("window").width * .01,
    paddingVertical: Dimensions.get("window").width * .01,
    backgroundColor: bases.base02,
    marginTop,
    textAlignVertical: "top",
    color: bases.base0,
    width,
  },
  recipientButtonGroup: {
    flexDirection: "row",
    justifyContent: "space-between",
    width: Dimensions.get("window").width * .6,
  },
  emailForm: {
    alignItems: "center",
    alignSelf: "center",
    backgroundColor: bases.base03,
    flex: 1,
    justifyContent: "flex-start",
    marginTop: Dimensions.get("window").height * .1,
  },
  root: {
    backgroundColor: bases.base03,
    flex: 1,
  },
  subject: {
    paddingHorizontal: Dimensions.get("window").width * .01,
    paddingVertical: Dimensions.get("window").width * .01,
    backgroundColor: bases.base02,
    color: bases.base0,
    marginTop,
    width,
  },
});
