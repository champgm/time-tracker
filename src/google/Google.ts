
import { Buffer } from "buffer";
import qs from "qs";
import { configuration } from "../Configuration";
import { AppendCellsRequest } from "./AppendCellsRequest";

export async function getAccessToken() {
  const parameters: RequestInit = {
    method: "POST",
    headers: {
      "Content-Type": "application/x-www-form-urlencoded",
    },
    body: qs.stringify({
      ...configuration.googleAuth.tokenRequestBody,
    }),
  };
  const fetchResult = await fetch(configuration.googleAuth.tokenEndpoint, parameters);
  return (await fetchResult.json()).access_token;
}

export function formatEmail(sender: string, recipient: string, subject: string, body: string) {
  return `To: ${recipient}
From: ${sender}
Subject: ${handleUnicode(subject)}

${body ? body : "."}\n`;
}

// https://github.com/googleapis/google-api-nodejs-client/issues/739
// tslint:disable-next-line:max-line-length
// https://github.com/googleapis/google-api-nodejs-client/pull/1088/files/7dd0d7770911688a5e6018d9418381be81253146#diff-ed4183b64bd56b93f40d1aaa8a7597d0
function handleUnicode(stringValue) {
  const buffer = Buffer.from(stringValue);
  const encoded = buffer.toString("base64");
  const full = `=?utf-8?B?${encoded}?=`;
  return full;
}


export async function addRow(date: Date, action: string, note: string) {
  const bearerTokenPromise = getAccessToken();
  const timestamp = date.getTime();
  const appendCellsRequest: AppendCellsRequest = {
    rows: [{
      values: [
        { userEnteredValue: { numberValue: timestamp } },
        { userEnteredValue: { stringValue: date.toLocaleString("en-US", { timeZone: "America/New_York" }) } },
        { userEnteredValue: { stringValue: action } },
        { userEnteredValue: { stringValue: note } },
      ],
    }],
    fields: '*',
    sheetId: configuration.spreadsheet.sheetId,
  }

  const authorizationHeader = `Bearer ${await bearerTokenPromise}`;
  const body = { requests: [{ appendCells: appendCellsRequest }] };
  const parameters: RequestInit = {
    method: "POST",
    headers: { authorization: authorizationHeader, },
    body: JSON.stringify(body),
  };

  const batchUpdateUrl = `${configuration.googleAuth.sheetsEndpoint}/${configuration.spreadsheet.spreadsheetId}:batchUpdate`
  console.log(`Updating sheet with URI: ${batchUpdateUrl}`);
  console.log(`Adding rows with parameters: ${JSON.stringify(parameters, null, 2)}`);
  console.log(`JSON body: ${JSON.stringify(body, null, 2)}`);

  const fetchResult = await fetch(batchUpdateUrl, parameters);
  console.log(`Row add result: ${JSON.stringify(fetchResult, null, 2)}`);
  const fetchResultJson = await fetchResult.json();
  console.log(`Row add result JSON: ${JSON.stringify(fetchResultJson, null, 2)}`);
}
