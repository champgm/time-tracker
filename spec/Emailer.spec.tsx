jest.mock("../src/Google");

import React from "react";
import { fireEvent, render, RenderAPI } from "react-native-testing-library";
import renderer from "react-test-renderer";

import { recipientConfiguration } from "../src/Configuration/Spreadsheet";
import { Emailer } from "../src/Emailer";
import { sendEmail } from "../src/google/Google";

describe("Emailer", () => {
  let emailer: RenderAPI;
  beforeEach(() => {
    (sendEmail as any).mockClear();
    emailer = render(<Emailer />);
  });

  it("renders correctly", () => {
    const tree = renderer.create(<Emailer />);
    const treeJson = tree.toJSON();
    expect(treeJson).toMatchSnapshot();
  });

  it("Should gather and send all fields", () => {
    const testSubject = "Test Subject";
    const testBody = "Test Body";

    const subject = emailer.getByTestId("cd7b46a4-2d81-47bf-abc1-7142aba8a7b0");
    fireEvent.changeText(subject, testSubject);

    const body = emailer.getByTestId("6b50a3ac-a102-4150-823c-e20d44f0c84d");
    fireEvent.changeText(body, testBody);

    const sendButton = emailer.getByTestId("2d8395f6-03a5-4c61-9c3b-595143aec8bf");
    fireEvent.press(sendButton);

    const defaultRecipient = recipientConfiguration.recipients
      .find((recipient) => recipient.name === recipientConfiguration.defaultRecipient);
    expect(sendEmail).toHaveBeenCalledWith(
      recipientConfiguration.sender,
      defaultRecipient.email,
      testSubject,
      testBody,
    );
  });

  it("Should be able to handle chinese text", () => {
    const testSubject = "我是一只猫";
    const testBody = "我有一个弟弟";

    const subject = emailer.getByTestId("cd7b46a4-2d81-47bf-abc1-7142aba8a7b0");
    fireEvent.changeText(subject, testSubject);

    const body = emailer.getByTestId("6b50a3ac-a102-4150-823c-e20d44f0c84d");
    fireEvent.changeText(body, testBody);

    const sendButton = emailer.getByTestId("2d8395f6-03a5-4c61-9c3b-595143aec8bf");
    fireEvent.press(sendButton);

    const defaultRecipient = recipientConfiguration.recipients
      .find((recipient) => recipient.name === recipientConfiguration.defaultRecipient);
    expect(sendEmail).toHaveBeenCalledWith(
      recipientConfiguration.sender,
      defaultRecipient.email,
      testSubject,
      testBody,
    );
  });

});
