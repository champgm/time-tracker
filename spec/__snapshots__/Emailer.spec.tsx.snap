// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Emailer renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "rgb(0,43,54)",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "rgb(0,43,54)",
        "flex": 1,
        "justifyContent": "flex-start",
        "marginTop": 133.4,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
          "width": 450,
        }
      }
    >
      <View
        accessible={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "transparent",
            "height": 60,
            "opacity": 0,
            "width": null,
            "zIndex": 10,
          }
        }
        testID="aws-btn-content-view"
      >
        <View
          style={
            Object {
              "backgroundColor": "rgba(0, 0, 0, 0.15)",
              "borderRadius": 4,
              "bottom": -2,
              "height": 56,
              "left": "1%",
              "position": "absolute",
              "transform": Array [
                Object {
                  "translateY": 0,
                },
              ],
              "width": "98%",
            }
          }
          testID="aws-btn-shadow"
        />
        <View
          style={
            Array [
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
              },
              Object {
                "backgroundColor": "rgb(0,43,54)",
                "borderRadius": 4,
                "height": 56,
                "width": null,
              },
            ]
          }
          testID="aws-btn-bottom"
        />
        <View
          style={
            Object {
              "borderRadius": 4,
              "height": 56,
              "left": 0,
              "position": "absolute",
              "top": 0,
              "transform": Array [
                Object {
                  "translateY": 0,
                },
              ],
              "width": null,
            }
          }
          testID="aws-btn-content"
        >
          <View
            onLayout={[Function]}
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "justifyContent": "center",
                  "overflow": "hidden",
                  "zIndex": 3,
                },
                Object {
                  "backgroundColor": "rgb(88,110,117)",
                  "borderColor": null,
                  "borderRadius": 4,
                  "borderWidth": 0,
                },
              ]
            }
            testID="aws-btn-text"
          >
            <View
              style={
                Object {
                  "backgroundColor": "rgb(7,54,66)",
                  "height": 56,
                  "left": 0,
                  "opacity": 0,
                  "position": "absolute",
                  "top": 0,
                  "width": null,
                }
              }
              testID="aws-btn-active-background"
            />
            <Text
              style={
                Object {
                  "color": "#FFFFFF",
                  "fontFamily": null,
                  "fontSize": 14,
                  "fontWeight": "bold",
                  "opacity": 1,
                  "paddingBottom": 0,
                  "paddingHorizontal": 12,
                  "paddingTop": 0,
                  "textAlign": "center",
                  "transform": Array [
                    Object {
                      "scale": 1,
                    },
                  ],
                  "zIndex": 10,
                }
              }
              testID="aws-btn-content-text"
            >
               Mac Work 
            </Text>
          </View>
        </View>
      </View>
      <View
        accessible={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "transparent",
            "height": 60,
            "opacity": 0,
            "width": null,
            "zIndex": 10,
          }
        }
        testID="aws-btn-content-view"
      >
        <View
          style={
            Object {
              "backgroundColor": "rgba(0, 0, 0, 0.15)",
              "borderRadius": 4,
              "bottom": -2,
              "height": 56,
              "left": "1%",
              "position": "absolute",
              "transform": Array [
                Object {
                  "translateY": 0,
                },
              ],
              "width": "98%",
            }
          }
          testID="aws-btn-shadow"
        />
        <View
          style={
            Array [
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
              },
              Object {
                "backgroundColor": "rgb(-50,72,147)",
                "borderRadius": 4,
                "height": 56,
                "width": null,
              },
            ]
          }
          testID="aws-btn-bottom"
        />
        <View
          style={
            Object {
              "borderRadius": 4,
              "height": 56,
              "left": 0,
              "position": "absolute",
              "top": 0,
              "transform": Array [
                Object {
                  "translateY": 0,
                },
              ],
              "width": null,
            }
          }
          testID="aws-btn-content"
        >
          <View
            onLayout={[Function]}
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "justifyContent": "center",
                  "overflow": "hidden",
                  "zIndex": 3,
                },
                Object {
                  "backgroundColor": "rgb(38,139,210)",
                  "borderColor": null,
                  "borderRadius": 4,
                  "borderWidth": 0,
                },
              ]
            }
            testID="aws-btn-text"
          >
            <View
              style={
                Object {
                  "backgroundColor": "rgb(-43,83,159)",
                  "height": 56,
                  "left": 0,
                  "opacity": 0,
                  "position": "absolute",
                  "top": 0,
                  "width": null,
                }
              }
              testID="aws-btn-active-background"
            />
            <Text
              style={
                Object {
                  "color": "#FFFFFF",
                  "fontFamily": null,
                  "fontSize": 14,
                  "fontWeight": "bold",
                  "opacity": 1,
                  "paddingBottom": 0,
                  "paddingHorizontal": 12,
                  "paddingTop": 0,
                  "textAlign": "center",
                  "transform": Array [
                    Object {
                      "scale": 1,
                    },
                  ],
                  "zIndex": 10,
                }
              }
              testID="aws-btn-content-text"
            >
               Mac 
            </Text>
          </View>
        </View>
      </View>
      <View
        accessible={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "transparent",
            "height": 60,
            "opacity": 0,
            "width": null,
            "zIndex": 10,
          }
        }
        testID="aws-btn-content-view"
      >
        <View
          style={
            Object {
              "backgroundColor": "rgba(0, 0, 0, 0.15)",
              "borderRadius": 4,
              "bottom": -2,
              "height": 56,
              "left": "1%",
              "position": "absolute",
              "transform": Array [
                Object {
                  "translateY": 0,
                },
              ],
              "width": "98%",
            }
          }
          testID="aws-btn-shadow"
        />
        <View
          style={
            Array [
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
              },
              Object {
                "backgroundColor": "rgb(0,43,54)",
                "borderRadius": 4,
                "height": 56,
                "width": null,
              },
            ]
          }
          testID="aws-btn-bottom"
        />
        <View
          style={
            Object {
              "borderRadius": 4,
              "height": 56,
              "left": 0,
              "position": "absolute",
              "top": 0,
              "transform": Array [
                Object {
                  "translateY": 0,
                },
              ],
              "width": null,
            }
          }
          testID="aws-btn-content"
        >
          <View
            onLayout={[Function]}
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "justifyContent": "center",
                  "overflow": "hidden",
                  "zIndex": 3,
                },
                Object {
                  "backgroundColor": "rgb(88,110,117)",
                  "borderColor": null,
                  "borderRadius": 4,
                  "borderWidth": 0,
                },
              ]
            }
            testID="aws-btn-text"
          >
            <View
              style={
                Object {
                  "backgroundColor": "rgb(7,54,66)",
                  "height": 56,
                  "left": 0,
                  "opacity": 0,
                  "position": "absolute",
                  "top": 0,
                  "width": null,
                }
              }
              testID="aws-btn-active-background"
            />
            <Text
              style={
                Object {
                  "color": "#FFFFFF",
                  "fontFamily": null,
                  "fontSize": 14,
                  "fontWeight": "bold",
                  "opacity": 1,
                  "paddingBottom": 0,
                  "paddingHorizontal": 12,
                  "paddingTop": 0,
                  "textAlign": "center",
                  "transform": Array [
                    Object {
                      "scale": 1,
                    },
                  ],
                  "zIndex": 10,
                }
              }
              testID="aws-btn-content-text"
            >
               Na 
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View>
      <TextInput
        allowFontScaling={true}
        autoFocus={true}
        onChangeText={[Function]}
        rejectResponderTermination={true}
        style={
          Object {
            "backgroundColor": "rgb(7,54,66)",
            "color": "rgb(131,148,150)",
            "marginTop": 40.019999999999996,
            "paddingHorizontal": 7.5,
            "paddingVertical": 7.5,
            "width": 525,
          }
        }
        testID="cd7b46a4-2d81-47bf-abc1-7142aba8a7b0"
        underlineColorAndroid="transparent"
      />
    </View>
    <View>
      <TextInput
        allowFontScaling={true}
        multiline={true}
        numberOfLines={8}
        onChangeText={[Function]}
        rejectResponderTermination={true}
        style={
          Object {
            "backgroundColor": "rgb(7,54,66)",
            "color": "rgb(131,148,150)",
            "marginTop": 40.019999999999996,
            "paddingHorizontal": 7.5,
            "paddingVertical": 7.5,
            "textAlignVertical": "top",
            "width": 525,
          }
        }
        testID="6b50a3ac-a102-4150-823c-e20d44f0c84d"
        underlineColorAndroid="transparent"
      />
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginTop": 40.019999999999996,
          "width": 525,
        }
      }
    >
      <View
        accessible={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "transparent",
            "height": 60,
            "opacity": 1,
            "width": 175,
            "zIndex": 10,
          }
        }
        testID="aws-btn-content-view"
      >
        <View
          style={
            Object {
              "backgroundColor": "rgba(0, 0, 0, 0.15)",
              "borderRadius": 4,
              "bottom": -2,
              "height": 56,
              "left": "1%",
              "position": "absolute",
              "transform": Array [
                Object {
                  "translateY": 0,
                },
              ],
              "width": "98%",
            }
          }
          testID="aws-btn-shadow"
        />
        <View
          style={
            Array [
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
              },
              Object {
                "backgroundColor": "rgb(132,-17,-16)",
                "borderRadius": 4,
                "height": 56,
                "width": 175,
              },
            ]
          }
          testID="aws-btn-bottom"
        />
        <View
          style={
            Object {
              "borderRadius": 4,
              "height": 56,
              "left": 0,
              "position": "absolute",
              "top": 0,
              "transform": Array [
                Object {
                  "translateY": 0,
                },
              ],
              "width": 175,
            }
          }
          testID="aws-btn-content"
        >
          <View
            onLayout={[Function]}
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "justifyContent": "center",
                  "overflow": "hidden",
                  "zIndex": 3,
                },
                Object {
                  "backgroundColor": "rgb(220,50,47)",
                  "borderColor": null,
                  "borderRadius": 4,
                  "borderWidth": 0,
                },
              ]
            }
            testID="aws-btn-text"
          >
            <View
              style={
                Object {
                  "backgroundColor": "rgb(139,-6,-4)",
                  "height": 56,
                  "left": 0,
                  "opacity": 0,
                  "position": "absolute",
                  "top": 0,
                  "width": 175,
                }
              }
              testID="aws-btn-active-background"
            />
            <Text
              style={
                Object {
                  "color": "#FFFFFF",
                  "fontFamily": null,
                  "fontSize": 14,
                  "fontWeight": "bold",
                  "opacity": 1,
                  "paddingBottom": 0,
                  "paddingHorizontal": 12,
                  "paddingTop": 0,
                  "textAlign": "center",
                  "transform": Array [
                    Object {
                      "scale": 1,
                    },
                  ],
                  "zIndex": 10,
                }
              }
              testID="aws-btn-content-text"
            >
              Clear
            </Text>
          </View>
        </View>
      </View>
      <View
        accessible={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "transparent",
            "height": 60,
            "opacity": 1,
            "width": 175,
            "zIndex": 10,
          }
        }
        testID="aws-btn-content-view"
      >
        <View
          style={
            Object {
              "backgroundColor": "rgba(0, 0, 0, 0.15)",
              "borderRadius": 4,
              "bottom": -2,
              "height": 56,
              "left": "1%",
              "position": "absolute",
              "transform": Array [
                Object {
                  "translateY": 0,
                },
              ],
              "width": "98%",
            }
          }
          testID="aws-btn-shadow"
        />
        <View
          style={
            Array [
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
              },
              Object {
                "backgroundColor": "rgb(45,86,-63)",
                "borderRadius": 4,
                "height": 56,
                "width": 175,
              },
            ]
          }
          testID="aws-btn-bottom"
        />
        <View
          style={
            Object {
              "borderRadius": 4,
              "height": 56,
              "left": 0,
              "position": "absolute",
              "top": 0,
              "transform": Array [
                Object {
                  "translateY": 0,
                },
              ],
              "width": 175,
            }
          }
          testID="aws-btn-content"
        >
          <View
            onLayout={[Function]}
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "justifyContent": "center",
                  "overflow": "hidden",
                  "zIndex": 3,
                },
                Object {
                  "backgroundColor": "rgb(133,153,0)",
                  "borderColor": null,
                  "borderRadius": 4,
                  "borderWidth": 0,
                },
              ]
            }
            testID="aws-btn-text"
          >
            <View
              style={
                Object {
                  "backgroundColor": "rgb(52,97,-51)",
                  "height": 56,
                  "left": 0,
                  "opacity": 0,
                  "position": "absolute",
                  "top": 0,
                  "width": 175,
                }
              }
              testID="aws-btn-active-background"
            />
            <Text
              style={
                Object {
                  "color": "#FFFFFF",
                  "fontFamily": null,
                  "fontSize": 14,
                  "fontWeight": "bold",
                  "opacity": 1,
                  "paddingBottom": 0,
                  "paddingHorizontal": 12,
                  "paddingTop": 0,
                  "textAlign": "center",
                  "transform": Array [
                    Object {
                      "scale": 1,
                    },
                  ],
                  "zIndex": 10,
                }
              }
              testID="aws-btn-content-text"
            >
              Send
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
